crates/oxc_linter/src/rules/eslint/max_classes_per_file.rs

```diff
@@ -0,0 +1,194 @@
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule};
+
+fn max_classes_per_file_diagnostic(x0: &str, span1: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warning(format!("eslint(max-classes-per-file): {x0:?}"))
+        .with_help("Reduce the number of classes in this file")
+        .with_labels([span1.into()])
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct MaxClassesPerFile(Box<MaxClassesPerFileConfig>);
+
+#[derive(Debug, Clone)]
+pub struct MaxClassesPerFileConfig {
+    pub max: usize,
+    pub ignore_expressions: bool,
+}
+
+impl std::ops::Deref for MaxClassesPerFile {
+    type Target = MaxClassesPerFileConfig;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Default for MaxClassesPerFileConfig {
+    fn default() -> Self {
+        Self { max: 1, ignore_expressions: false }
+    }
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Enforce a maximum number of classes per file
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Files containing multiple classes can often result in a less navigable and poorly
+    /// structured codebase. Best practice is to keep each file limited to a single responsibility.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// class Foo {}
+    /// class Bar {}
+    /// ```
+    MaxClassesPerFile,
+    correctness,
+);
+
+impl Rule for MaxClassesPerFile {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let config = value.get(0);
+        if let Some(max) = config
+            .and_then(Value::as_number)
+            .and_then(serde_json::Number::as_u64)
+            .and_then(|v| usize::try_from(v).ok())
+        {
+            Self(Box::new(MaxClassesPerFileConfig { max, ignore_expressions: false }))
+        } else {
+            let max = value
+                .get(0)
+                .and_then(|config| config.get("max"))
+                .and_then(serde_json::Value::as_number)
+                .and_then(serde_json::Number::as_u64)
+                .map_or(1, |v| usize::try_from(v).unwrap_or(1));
+
+            let ignore_expressions = value
+                .get(0)
+                .and_then(|config| config.get("ignoreExpressions"))
+                .and_then(serde_json::Value::as_bool)
+                .unwrap_or(false);
+            Self(Box::new(MaxClassesPerFileConfig { max, ignore_expressions }))
+        }
+    }
+
+    fn run_once(&self, ctx: &LintContext<'_>) {
+        let mut class_count = 0;
+        let mut span = Span::new(0, 0);
+
+        ctx.semantic().classes().iter_enumerated().for_each(|(_class_id, node_id)| {
+            let kind = ctx.nodes().kind(*node_id);
+
+            if class_count == self.max {
+                if let AstKind::Class(class) = kind {
+                    span = class.span;
+                }
+            }
+
+            if !kind.is_declaration() && self.ignore_expressions {
+            } else {
+                class_count += 1;
+            }
```

Comment:
Maybe there is a more Rust-like way to write this?

Reply:
I could change this into `filter().cnt()` and then look up `span` via `max` but I am not 100% sure how one does it as `ClassId` is not an integer but a Struct ClassId?

Reply:
I think you can implement both `ignore_expression` is true or not logic separately

You can add the `len` method in classes and check `class_count <= self.max` first

```rust
let mut class_count = ctx.semantic.classes().len()
if class_count <= self.max {
   return;
}
```

If true
```rust
let ast_node_id = ctx.semantic().classes().get_node_id(self.max as ClassId)
let AstKind::Class(class) = ctx.nodes().kind(*node_id);
span = class.span;
``` 

If false, you can use `ctx.semantic().classes().iter_enumerated().filter()` to get `class_count` and first `span`. if `class_count <= self.max` we can return

Reply:
Thanks for the suggestions, I implemented something along the same lines.